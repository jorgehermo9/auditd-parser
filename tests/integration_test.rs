use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
