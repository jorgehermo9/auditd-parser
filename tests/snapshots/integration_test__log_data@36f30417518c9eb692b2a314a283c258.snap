---
source: tests/integration_test.rs
expression: record
info: "type=SYSCALL msg=audit(1490801406.273:512226): arch=c000003e syscall=42 success=no exit=-115 a0=6 a1=7ffeb50e4570 a2=10 a3=fe items=0 ppid=1 pid=1170 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=\"google_ip_forwa\" exe=2F7573722F62696E2F707974686F6E322E373B3538643163636662202864656C6574656429 subj=system_u:system_r:unconfined_service_t:s0 key=(null)"
---
{
  "record_type": "SYSCALL",
  "timestamp": 1490801406273,
  "id": 512226,
  "fields": {
    "a0": "6",
    "a1": "7ffeb50e4570",
    "a2": "10",
    "a3": "fe",
    "arch": "c000003e",
    "auid": null,
    "comm": "google_ip_forwa",
    "egid": "root",
    "euid": "root",
    "exe": "/usr/bin/python2.7;58d1ccfb (deleted)",
    "exit": -115,
    "fsgid": "root",
    "fsuid": "root",
    "gid": "root",
    "items": "0",
    "key": "(null)",
    "pid": "1170",
    "ppid": "1",
    "ses": "4294967295",
    "sgid": "root",
    "subj": {
      "level": {
        "sensitivity": "s0"
      },
      "module": "SELinux",
      "role": "system_r",
      "type": "unconfined_service_t",
      "user": "system_u"
    },
    "success": false,
    "suid": "root",
    "syscall": "42",
    "tty": "(none)",
    "uid": "root"
  }
}
